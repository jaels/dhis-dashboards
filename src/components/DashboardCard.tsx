import React, { useState, Fragment } from 'react';

import { DashboardListItem, FullItemInfo, DashboardItem } from '../types/apiData'

/* @ts-ignore */
import { Card } from '@dhis2-ui/card'
import classnames from "classnames"

interface Props {
  expanded: boolean;
  itemInfo: DashboardListItem;
  expandedDetails: FullItemInfo | null | undefined;
  onExpandCardClick: (id: string) => void;
  currentFilters: string[]
}

const DashboardCard: React.FC<Props> = props => {
  const { expanded, itemInfo, expandedDetails, onExpandCardClick, currentFilters } = props;

  const [starred, setStarred] = useState<boolean>(localStorage.getItem(itemInfo.id) ? true : false)
  
  /* I would definitely combine the 2 functions below in one that returns an object with text and icon as keys
  but ts kept complaining in the render that the object could be undefined etc, and insisted, no matter what I did,
  so I left it with those 2 functions */

  const findItemsRelevantText = (item: DashboardItem) => {
    switch(item.type.toLowerCase()) {
      case "visualization":
        if(item.visualization)
        return item.visualization.name
        break;
      case "map":
        if(item.map)
        return item.map.name
        break;
      case "text":
        if(item.text)
        return item.text
        break;
      default: 
      return null
    }
  }
  
  const findItemsRelevantIcon = (item: DashboardItem) => {
    switch(item.type.toLowerCase()) {
      case "visualization":
        if(item.visualization)
        return "signal_cellular_alt"
        break;
      case "map":
        if(item.map)
        return "public"
        break;
      case "text":
        if(item.text)
        return "font_download"
        break;
      default: 
      return null
    }
  }

  const checkIfItemPassesFilters = (item: DashboardItem): boolean => {
    if(currentFilters.length) {
      if(currentFilters.indexOf(item.type.toLowerCase()) === -1)
      return false
    }
    return true
  }

  const handleStarClick = (id: string) => {
    if(localStorage.getItem(id)) {
      localStorage.removeItem(id)
      setStarred(false)
    }
    else {
      localStorage.setItem(id, "true")
      setStarred(true)
    }
  }

  return (
    <Card className="card-container">
      <div className="collapsed-card">
        <p>{itemInfo.displayName}</p>
        <div className="icons-area">
          <span 
            className={classnames("material-icons", "button-icons", "star-icon")} 
            onClick={() => handleStarClick(itemInfo.id)}
          >
            {starred ? "star" : "star_outline"}
          </span>
          <span 
            data-testid="expand-arrow"
            className={classnames("material-icons", "button-icons", "arrow-icon")} 
            onClick={() => onExpandCardClick(itemInfo.id)}
          >
            {expanded ? "keyboard_arrow_up" : "keyboard_arrow_down"}
          </span>
        </div>
      </div>
      {expanded && expandedDetails && (
        <div className="expanded-card">
          {expandedDetails.dashboardItems.map((item, i) => (
            <Fragment key={i}>
              {findItemsRelevantText(item) && checkIfItemPassesFilters(item) &&
              <>
              <div className="dashboard-item-info">
                <span className="material-icons">{findItemsRelevantIcon(item)}</span>
                <span className="dashboard-item-text">{findItemsRelevantText(item)}</span>
              </div>
                {i !== expandedDetails.dashboardItems.length - 1 && <hr></hr>}
                </>
              }
            </Fragment>
          ))}
        </div>
      )}
    </Card>
  )
}

export default React.memo(DashboardCard)