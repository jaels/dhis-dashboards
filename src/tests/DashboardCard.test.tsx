import DashboardCard from '../components/DashboardCard';
import { render, screen, fireEvent } from '@testing-library/react';
import { mockExpanded } from './mockData'

describe ('DashboardCard', () => {
  const props = {
    expanded: true,
    itemInfo: {
      "displayName": "Antenatal Care",
      "id": "nghVC4wtyzi",
      "starred": true
    },
    expandedDetails: mockExpanded,
    onExpandCardClick: jest.fn(),
    currentFilters: []
  }

  it("renders the innitial info correctly", () => {
     /* @ts-ignore */
     const { asFragment } = render(<DashboardCard {...props} />)
     expect(asFragment()).toMatchSnapshot()   
     expect (screen.getByText("Antenatal Care")).toBeVisible()
  })

  it("renders the dashboard items of the first card", () => {
     /* @ts-ignore */
    render(<DashboardCard {...props} />);
    expect (screen.getByText("first dashboard item text")).toBeVisible()
    expect (screen.getByText("second dashboard item text")).toBeVisible()
  })

  it("renders the expand arrow and evokes the click function", () => {
     /* @ts-ignore */
    render(<DashboardCard {...props} />);
    expect(screen.getAllByTestId('expand-arrow')).toBeTruthy();

    const expandButton = screen.getAllByTestId('expand-arrow')[0];
    fireEvent.click(expandButton)
    expect(props.onExpandCardClick).toHaveBeenCalledTimes(1)
  })
})

  