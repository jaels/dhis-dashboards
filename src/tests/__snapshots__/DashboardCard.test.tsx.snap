// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardCard renders the innitial info correctly 1`] = `
<DocumentFragment>
  <div
    class="jsx-574779348 card-container"
    data-test="dhis2-uicore-card"
  >
    <div
      class="collapsed-card"
    >
      <p>
        Antenatal Care
      </p>
      <div
        class="icons-area"
      >
        <span
          class="material-icons button-icons star-icon"
        >
          star_outline
        </span>
        <span
          class="material-icons button-icons arrow-icon"
          data-testid="expand-arrow"
        >
          keyboard_arrow_up
        </span>
      </div>
    </div>
    <div
      class="expanded-card"
    >
      <div
        class="dashboard-item-info"
      >
        <span
          class="material-icons"
        >
          signal_cellular_alt
        </span>
        <span
          class="dashboard-item-text"
        >
          first dashboard item text
        </span>
      </div>
      <hr />
      <div
        class="dashboard-item-info"
      >
        <span
          class="material-icons"
        >
          public
        </span>
        <span
          class="dashboard-item-text"
        >
          second dashboard item text
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
